dbms/src/Storages/Transaction/Collator.h

```diff
@@ -91,4 +119,119 @@ extern TiDBCollators dummy_collators;
 extern std::vector<std::string> dummy_sort_key_contaners;
 extern std::string dummy_sort_key_contaner;
 
+struct TiDBCollatorPtrImpl : TiDBCollatorID
+{
+    class ITiDBCollator const * ptr{};
+
+    TiDBCollatorPtrImpl(class ITiDBCollator const * ptr_, int32_t collator_id_)
+        : TiDBCollatorID(collator_id_)
+        , ptr(ptr_)
+    {
+    }
+
+    explicit TiDBCollatorPtrImpl(class ITiDBCollator const * ptr_)
+        : TiDBCollatorID(ptr_ ? ptr_->getCollatorId() : 0)
+        , ptr(ptr_)
+    {
+    }
+
+    ALWAYS_INLINE inline int compare(const char * s1, size_t length1, const char * s2, size_t length2) const
+    {
+        if (likely(canUseFastPath()))
```

Comment:
Will this `likely` hurt the performance of non-utf8mb4_bin collator?

Reply:
I think it will not hurt performance because the cost of local branch-jump is lower than virtual call.
`likely` is a **hint** for compiler to put branch closer to current instrument.

for example, with likely, maybe:
```
mov %eax,%esi // result = canUseFastPath()
test  %eax,%eax
je    ..... // jump to the branch DO NOT use fast path
.... // fast path branch
```